### PR TITLE
service layer

### PR DESCRIPTION
**Pull request by karpovdim**

### CODE REVIEW

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        return giftCertificateDtoList;
+    }
+
+    @Override
+    public List<GiftCertificateDto> getAllWithSortingAndFiltering(List<String> sortColumns,
+                                                                  List<String> orderType,
+                                                                  List<String> filterBy) {
+        List<GiftCertificateDto> giftCertificateDtoList = new ArrayList<>();
+        if (sortColumns == null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move this check to a separate method

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        return giftCertificateDtoList;
+    }
+
+    @Transactional
+    @Override
+    public GiftCertificateDto updateById(Long giftCertificateId, GiftCertificateDto giftCertificateDto) {
+        GiftCertificate giftCertificate = giftCertificateDto.getGiftCertificate();
+        List<Tag> tagList = giftCertificateDto.getCertificateTags();
+        if (tagList == null || giftCertificate == null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        isPriceValidForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        isDurationValidForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
+        return updateInfo;
+    }
+
+    private String getGiftCertificateName(GiftCertificate giftCertificate) {
+        String giftCertificateName = giftCertificate.getName();
+        boolean isCertificateExist = giftCertificateDao.getByName(giftCertificateName).isPresent();
+        if (isCertificateExist) {
+            throw new DuplicateEntityException("message.giftCertificateExists");
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        return giftCertificateName;
+    }
+
+    private void isDurationValidForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        Integer duration = giftCertificate.getDuration();
+        if (!giftCertificateValidator.isDurationValid(duration)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            throw new InvalidEntityException("message.cantUpdateInvalidDuration");
+        }
+        updateInfo.put("duration", duration);
+    }
+
+    private void isPriceValidForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        BigDecimal price = giftCertificate.getPrice();
+        if (price != null) {
+            if (!giftCertificateValidator.isPriceValid(price)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+                throw new InvalidEntityException("message.cantUpdateInvalidPrice");
+            }
+            updateInfo.put("price", price);
+        }
+    }
+
+    private void isValidDescriptionForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        String description = giftCertificate.getDescription();
+        if (description != null) {
+            if (!giftCertificateValidator.isDescriptionValid(description)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                throw new InvalidEntityException("message.cantUpdateInvalidDescription");
+            }
+            updateInfo.put("description", description);
+        }
+    }
+
+    private void isValidNameForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        String name = giftCertificate.getName();
+        if (name != null) {
+            if (!giftCertificateValidator.isNameValid(name)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            if (!giftCertificateValidator.isDescriptionValid(description)) {
+                throw new InvalidEntityException("message.cantUpdateInvalidDescription");
+            }
+            updateInfo.put("description", description);
+        }
+    }
+
+    private void isValidNameForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        String name = giftCertificate.getName();
+        if (name != null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: check for blank - use apache's library

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            if (!giftCertificateValidator.isPriceValid(price)) {
+                throw new InvalidEntityException("message.cantUpdateInvalidPrice");
+            }
+            updateInfo.put("price", price);
+        }
+    }
+
+    private void isValidDescriptionForUpdate(GiftCertificate giftCertificate, Map<String, Object> updateInfo, GiftCertificateValidatorImpl giftCertificateValidator) {
+        String description = giftCertificate.getDescription();
+        if (description != null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: check for blank - use apache's library

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+            if (!giftCertificateValidator.isNameValid(name)) {
+                throw new InvalidEntityException("message.cantUpdateInvalidName");
+            }
+            updateInfo.put("name", name);
+        }
+    }
+
+    private void isPresentGiftCertificate(Long id) {
+        Optional<GiftCertificate> giftCertificateOptional = giftCertificateDao.getById(id);
+        if (giftCertificateOptional.isEmpty()) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: Move it validator

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+            throw new NoSuchEntityException("message.notFound");
+        }
+    }
+
+    private void updateCertificateTags(List<Tag> tagList, Long giftCertificateId) {
+        for (Tag tag : tagList) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:193]: split logic to separate methods

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        return tagDao.getByName(tag.getName()).orElseThrow(() -> new NoSuchEntityException("message.notFound"));
+    }
+
+    private void validateGiftCertificate(GiftCertificate giftCertificate) {
+        if (!giftCertificateValidator.isValid(giftCertificate)) {
+            throw new InvalidEntityException("message.giftCertificateInvalid");
+        }
+    }
+
+    private void validateTags(List<Tag> tags) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: seems class for  GiftCertificate))

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    public List<Tag> getAll() {
+        return tagDao.getAll();
+    }
+
+    @Transactional
+    @Override
+    public void create(Tag tag) {
+        if (!tagValidator.isValid(tag)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: you should have methods like "validate"
No exceptions here  - move them to validate methods

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Transactional
+    @Override
+    public void deleteById(Long id) {
+        isPresent(id);
+        tagDao.deleteById(id);
+    }
+
+    private void isPresent(Long id) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: incorrect name  - when "is....." then boolean, else validate

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        if (tagDao.getById(id).isEmpty()) {
+            throw new NoSuchEntityException("message.tagDoesntExist");
+        }
+    }
+
+    private void isExists(Tag tag) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: incorrect name - when "is....." then boolean, else validate

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 3 above)
Comment (karpovdim) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: can i do it like this?
private String getGiftCertificateName(GiftCertificate giftCertificate) {
        String giftCertificateName = giftCertificate.getName();
        giftCertificateDao.getByName(giftCertificateName).orElseThrow(()-> new DuplicateEntityException("message.giftCertificateExists"));
        return giftCertificateName;
    }

Comment (KseniyaYudzichavaTR) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: yes

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        isPresentGiftCertificate(giftCertificateId);
+        giftCertificateDao.updateById(giftCertificateId, updateInfo(giftCertificate));
+        updateCertificateTags(tagList, giftCertificateId);
+        return buildGiftCertificateDto(giftCertificateDao.getById(giftCertificateId).orElseThrow(() -> new NoSuchEntityException("message.notFound")));
+    }
+
+    private Map<String, Object> updateInfo(GiftCertificate giftCertificate) {
+        Map<String, Object> updateInfo = new HashMap<>();
+        isValidNameForUpdate(giftCertificate, updateInfo, giftCertificateValidator);
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: You are changing the incoming parameter - it is forbidden

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+        giftCertificateDao.getByName(giftCertificateName).orElseThrow(
+                () -> new DuplicateEntityException(GIFT_CERTIFICATE_EXISTS_MSG));
+        return giftCertificateName;
+    }
+
+
+    private void updateCertificateTags(List<Tag> tagList, Long giftCertificateId) {
+        for (Tag tag : tagList) {
+            String tagName = tag.getName();
+            if (StringUtils.isNotEmpty(tagName)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:195]: split method

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        tagDao.create(tag);
+    }
+
+    @Transactional
+    @Override
+    public void updateNameById(Long id, Tag tag) {
+        tagValidator.checkPresenceTagById(id,tagDao);
+        if (StringUtils.isNotEmpty(tag.getName())) {
+            tagDao.updateNameById(id, tag.getName());
+        } else {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: move to validate method

File: service/src/main/java/com/epam/esm/validator/impl/TagValidatorImpl.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+public class TagValidatorImpl implements Validator<Tag> {
+    @Value("${tag.maxNameLength}")
+    private int maxNameLength;
+    @Value("${tag.minNameLength}")
+    private int minNameLength;
+
+    @Override
+    public boolean isValid(Tag tag) {
+        String tagName = tag.getName();
+        if (tagName != null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/validator/impl/TagValidatorImpl.java]: simplify the logic

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java (hunk 16 above)
Comment (karpovdim) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: these are old changes, i fixed them. and I don't know why they are visible

Comment (karpovdim) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: private Map<String, Object> updateInfo(GiftCertificate giftCertificate) {
        Map<String, Object> updateInfo = new HashMap<>();
        if (giftCertificateValidator.isValidNameForUpdate(giftCertificate)) {
            updateInfo.put(GIFT_CERTIFICATE_NAME.getValue(), giftCertificate.getName());
        }
        if (giftCertificateValidator.isValidDescriptionForUpdate(giftCertificate)) {
            updateInfo.put(GIFT_CERTIFICATE_DESCRIPTION.getValue(), giftCertificate.getDescription());
        }
        if (giftCertificateValidator.isPriceValidForUpdate(giftCertificate)) {
            updateInfo.put(GIFT_CERTIFICATE_PRICE.getValue(), giftCertificate.getPrice());
        }
        if (giftCertificateValidator.isDurationValidForUpdate(giftCertificate)) {
            updateInfo.put(GIFT_CERTIFICATE_DURATION.getValue(), giftCertificate.getDuration());
        }
        return updateInfo;
    }

Comment (karpovdim) [reply service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: is it ok?

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+        this.giftCertificateDao = giftCertificateDao;
+        this.tagValidator = tagValidator;
+        this.tagDao = tagDao;
+    }
+
+    @Override
+    public List<GiftCertificateDto> getRoute(String tagName, List<String> sortColumns,
+                                             List<String> orderType, List<String> filterBy,
+                                             Long giftCertificateId, String allWithTags) {
+        if (giftCertificateId != null) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: too many return statements

File: service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+        giftCertificateDao.updateById(giftCertificateId, updateInfo(giftCertificate));
+        updateCertificateTags(tagList, giftCertificateId);
+        return buildGiftCertificateDto(giftCertificateDao.getById(giftCertificateId).orElseThrow(
+                () -> new NoSuchEntityException(NOT_FOUND_MSG)));
+    }
+
+
+    private Map<String, Object> updateInfo(GiftCertificate giftCertificate) {
+        Map<String, Object> updateInfo = new HashMap<>();
+        if (giftCertificateValidator.isValidNameForUpdate(giftCertificate)) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java]: too many ifs

File: service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    @Autowired
+    public TagServiceImpl(TagDao tagDao, TagValidatorImpl tagValidator) {
+        this.tagDao = tagDao;
+        this.tagValidator = tagValidator;
+    }
+
+    @Override
+    public List<Tag> getRoute(Long tagId, String tagName, Long giftCertificateId) {
+        if (tagId != null) {
+            return getById(tagId);
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: too many "return"

File: service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+        }
+    }
+
+    public void checkParamsNotNull(GiftCertificate giftCertificate, List<Tag> tagList) {
+        if (isListNull(tagList) || isNull(giftCertificate)) {
+            throw new InvalidEntityException(CANT_UPDATE_GIFT_CERTIFICATE);
+        }
+    }
+
+    public boolean isDurationValidForUpdate(GiftCertificate giftCertificate) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java:92]: no sense in boolean method - it never returns false

File: service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+        Integer duration = giftCertificate.getDuration();
+        if (!isDurationValid(duration)) {
+            throw new InvalidEntityException(CANT_UPDATE_INVALID_DURATION_MSG);
+        }
+        return true;
+    }
+
+    public boolean isPriceValidForUpdate(GiftCertificate giftCertificate) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java:100]: no sense in boolean method - it never returns false

File: service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+        BigDecimal price = giftCertificate.getPrice();
+        if (price != null) {
+            if (!isPriceValid(price)) {
+                throw new InvalidEntityException(CANT_UPDATE_INVALID_PRICE_MSG);
+            }
+        }
+        return true;
+    }
+
+    public boolean isValidDescriptionForUpdate(GiftCertificate giftCertificate) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java:110]: no sense in boolean method - it never returns false

File: service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+        String description = giftCertificate.getDescription();
+        if (StringUtils.isNotBlank(description)) {
+            if (!isDescriptionValid(description)) {
+                throw new InvalidEntityException(CANT_UPDATE_INVALID_DESCRIPTION_MSG);
+            }
+        }
+        return true;
+    }
+
+    public boolean isValidNameForUpdate(GiftCertificate giftCertificate) {
```
Comment (KseniyaYudzichavaTR) [inline service/src/main/java/com/epam/esm/validator/impl/GiftCertificateValidatorImpl.java:120]: no sense in boolean method - it never returns false

**Review by KseniyaYudzichavaTR — APPROVED** [review]

**Pull request merged**